API_HTTP.md

```diff
@@ -1,21 +1,598 @@
+
+The only exception to this is streaming responses where errors may be added to the output stream.
+
+If an error is added to the output stream, a final multipart boundary will be added and then the response will end and no more data will be sent.
+
+Example:
+
+> An error during a non-streaming resource response
+
+```
+content-type: application/ipfs-error-v1; charset=UTF-8
+
+{ "message": "An unknown error occured", "code": "ERR_UNKNOWN_ERROR", "stack": [ "line1", "line2", "line3" ] }
+```
+
+Example:
+
+> An error during a stream
+
+```
+content-type: multipart/mixed; boundary=boundary
+
+--boundary
+content-type: application/ipfs-unixfs-v1-file
+
+{ "path": "/foo", "cid": "bafyfoo"' }
+--boundary
+content-type: application/ipfs-error-v1; charset=UTF-8
+
+{ "message": "An unknown error occured", "code": "ERR_UNKNOWN_ERROR", "stack": [ "line1", "line2", "line3" ] }
+--boundary
+```
+
+#### Data types
+
+ * `application/ipfs-error-v1`
+
+```
+{
+  code: String  // an error code
+  message: String  // an optional description of the error
+  stack: Array[String]  // an optional stack trace of where the error occured
+}
+```
+
+### HTTP Verbs
+
+Supported HTTP verbs are interpreted to mean the following:
+
+* `GET` Retrieve a resource
+* `POST` Create a resource
+* `PUT` Replace a resource
+* `PATCH` Update a resource by sending a partial resource
+* `DELETE` Remove a resource
+* `OPTIONS` Retrieve a list of verbs applicable to a given resource
+* `HEAD` Retrieve resource metadata
+
+### Resource names
+
+Where a resource is one of many, they are specified using plural names and an identifier (unless creating a new resource).
+
+E.g.
+
+```
+POST /files
+GET /files/bafyFile
+```
+
+Where a resource is a singleton, it is specified using singular names.  Creating new singleton resources is disallowed and a `403 Forbidden` response must be returned.
+
+E.g.
+
+```
+GET /config
+PATCH /config
+
+POST /config // returns 403 Forbidden
+```
+
+### Response codes
+
+If a request cannot be processed an appropriate HTTP status code must be returned.
+
+If the fault is on the part of the client (e.g. parameter validation failed or the request was otherwise invalid, etc), a `4xx` series status should be returned.
+
+If the fault is on the part of the API (e.g. a network failure or an error condition due to misconfiguration, etc) a `5xx` series status should be returned.
+
+In some cases specific status codes will indicate certain request outcomes and must be returned, these will be noted in the `Resources` section below.
+
+## Resources
+
+### `/api/v1/files`
+
+The `files` resource allows interactions with IPFS filesystems.
+
+All `files` operations take a path which can be an IPFS path (e.g. `/ipfs/bafyFile` or `/ipfs/bafyDir/file.txt`) or an [MFS] path (e.g. `/path/to/file`).
+
+Where conflicts arise, the IPFS path takes priority.
+
+#### Data types
+
+ * `application/ipfs-unixfs-v1-file`
```

Comment:
Completely subjective.  I prefer concise kebab notation over dots because weird over specified mime types  like `application/vnd.ms-word.document.macroEnabled.12` give me the twitchy eye.